ddev/src/ddev/cli/meta/scripts/upgrade_python.py

```diff
@@ -20,57 +21,200 @@ def upgrade_python(app: Application, version: str):
+    constant_file = app.repo.path / 'ddev' / 'src' / 'ddev' / 'repo' / 'constants.py'
+
+    lines = constant_file.read_text().splitlines(keepends=True)
+    for i, line in enumerate(lines):
+        if line.startswith('PYTHON_VERSION = '):
+            lines[i] = line.replace(old_version, new_version)
+            break
+
+    constant_file.write_text(''.join(lines))
+    tracker.success()
+
+
+def update_pyproject_file(target, new_version: str, old_version: str, tracker: ValidationTracker):
+    import tomlkit
+
+    config_file = target.path / 'pyproject.toml'
+    config = tomlkit.parse(config_file.read_text())
+    changed = False
+
+    classifiers = config.get('project', {}).get('classifiers', [])
+    for index, classifier in enumerate(classifiers):
+        if classifier == f"Programming Language :: Python :: {old_version}":
+            classifiers[index] = f"Programming Language :: Python :: {new_version}"
+            changed = True
+            tracker.success()
+            break
+
+    if changed:
+        config_file.write_text(tomlkit.dumps(config))
+
+
+def update_hatch_file(app: Application, target_path, new_version: str, old_version: str, tracker: ValidationTracker):
+    import tomlkit
+
+    config_file = target_path / 'hatch.toml'
+    test_config = tomlkit.parse(config_file.read_text())
+    changed = False
+
+    for env in test_config.get('envs', {}).values():
+        if update_hatch_env(app, env, new_version, old_version, config_file, tracker):
+            changed = True
+
+    if changed:
+        config_file.write_text(tomlkit.dumps(test_config))
+
+
+def update_hatch_env(
+    app: Application, env, new_version: str, old_version: str, config_file, tracker: ValidationTracker
+) -> bool:
+    changed = False
+
+    default_python = env.get('python', '')
+    if default_python == old_version:
+        env['python'] = new_version
         tracker.success()
+        changed = True
 
-    tracker.display()
+    for variables in env.get('matrix', []):
+        pythons = variables.get('python', [])
+        for i, python in enumerate(pythons):
+            if python == old_version:
+                pythons[i] = new_version
+                tracker.success()
+                changed = True
 
-    if tracker.errors:  # no cov
-        app.abort()
+    for overrides in env.get('overrides', {}).get('matrix', {}).get('python', {}).values():
+        for override in overrides:
+            pythons = override.get('if', [])
+            for i, python in enumerate(pythons):
+                if python == old_version:
+                    pythons[i] = new_version
+                    tracker.success()
+                    changed = True
+
+    if isinstance(env.get('overrides', {}), dict):
+        for name in list(env.get('overrides', {}).get('name', {}).keys()):
+            if f"py{old_version}" in name:
+                # TODO I don't find a way to keep the exact same format when I modify this.
```

Comment:
I tried: 

```
    for name in list(overrides['name'].keys()):
        if f"py{old_version}" in name:
            overrides['name'][name.replace(f"py{old_version}", f"py{new_version}")] = overrides['name'].pop(name)
            tracker.success()
            changed = True
```

Reply:
I double checked, the generated dict is exactly the same as the one we read from the file, but the output file is different (using tomlkit, I'd like to avoid search/replace as much as possible)

Reply:
fair enough, unfortunately tomlkit may be less reliable in this case :( If you hesitate with search/replace, we can punt on this then. If you have time might be worth opening a bug report to tomlkit.